Refactor to generalize services

darinkrauss opened the pull request:
- Replace OSLog with DiagnosticLog
- Move GlucoseThreshold, LoopSettings, and TempBasalRecommendation from Loop
- Move BatteryStatus, IOBStatus, and PumpStatus from rileylink_ios
- Add Analytics, Logging, and RemoteData protocols
- Add Service and ServiceUI protocols
- Implement common ServiceTableViewController and ServiceViewController for simple services
- Implement MockService to record events from Analytics, Logging, and RemoteData procotols

[This is not up-to-date with the `dev` branch. I will include that next.]

darinkrauss committed:
Refactor to generalize services

- Replace OSLog with DiagnosticLog
- Move GlucoseThreshold, LoopSettings, and TempBasalRecommendation from Loop
- Move BatteryStatus, IOBStatus, and PumpStatus from rileylink_ios
- Add Analytics, Logging, and RemoteData protocols
- Add Service and ServiceUI protocols
- Implement common ServiceTableViewController and ServiceViewController for simple services
- Implement MockService to record events from Analytics, Logging, and RemoteData procotols

darinkrauss commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Design new Service and ServiceUI protocols](https://trello.com/c/4hlXSYN6/695-design-new-service-and-serviceui-protocols)

darinkrauss commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Design new RemoteData, Logging, and Analytics protocols](https://trello.com/c/gvq8JOwr/694-design-new-remotedata-logging-and-analytics-protocols)

darinkrauss commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Refactor existing services to support new Service and ServiceUI protocols](https://trello.com/c/rUjClkL2/611-refactor-existing-services-to-support-new-service-and-serviceui-protocols)

darinkrauss commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Refactor service settings UI to support Service and ServiceUI](https://trello.com/c/XoV3ZGlJ/703-refactor-service-settings-ui-to-support-service-and-serviceui)

darinkrauss commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Refactor existing services to support new RemoteData, Logging, and Analytics protocols](https://trello.com/c/0jXmC3cG/704-refactor-existing-services-to-support-new-remotedata-logging-and-analytics-protocols)

ps2 commented on LoopKit/LoopSettings.swift:
```diff
@@ -0,0 +1,108 @@
+//
+
+
+public struct LoopSettings: Equatable {
```
This schema is a Loop specific thing, and shouldn't be in LoopKit. The Service API could provide a general "settings" store facility, perhaps key/value based. If the latter, some batching mechanism should be considered.

darinkrauss replied:
This is fixed. It was pulled in as needed by the Analytics protocol. With the update to the Analytics protocol to be more generic, this struct is no longer required in LoopKit and has been moved back to Loop.

ps2 commented on LoopKit/InsulinKit/TempBasalRecommendation.swift:
```diff
@@ -0,0 +1,69 @@
+}
+
+
+extension TempBasalRecommendation {
```
The functions in this extension are probably better kept in DoseMath as they are part of the Loop dosing algorithm.

darinkrauss replied:
Fixed. Extension is moved into DoseMath.

ps2 commented on LoopKit/Analytics.swift:
```diff
@@ -0,0 +1,62 @@
+
+    // MARK: - UIApplicationDelegate
+
+    func application(didFinishLaunchingWithOptions launchOptions: [AnyHashable: Any]?)
```
This set of event types is a Loop app level concern, and probably shouldn't be in LoopKit.  If Service protocol will expose an analytics API, it probably needs to be generic to handle the changing set of event types.

darinkrauss replied:
Fixed. The Analytics protocol has been made more general and the Loop-level analytics functions are fully contained within Loop.

ps2 commented on LoopKit/PumpStatus.swift:
```diff
@@ -0,0 +1,30 @@
+//
+
+
+public struct PumpStatus {
```
This part of the upload varies by pump. I'm wondering if we want an opaque Codable object for this.  The use cases it needs to support are 1) the hud widgets in Loop Follow, and 2) existing 'pills' in NightScout.

Some of the information here, like pump IOB, is not needed (remote monitoring tools should be displaying Loop IOB instead).  And some info, like `bolusing` and `suspended` should be covered by the other stream of dose information.

darinkrauss replied:
This struct (and two others) were erroneously moved into LoopKit from rileylink_ios. They are now moved back to rileylink_ios.

ps2 commented on LoopKit/Service.swift:
```diff
@@ -0,0 +1,101 @@
+
+public protocol ServiceDelegate: AnyObject {
+
+    /// Informs the delegate that the specified service was updated.
```
What aspect of the service updated?  i.e. what would the delegate do in response to this message?

darinkrauss replied:
Updated protocol documentation with further detail.

ps2 commented on LoopKit/Service.swift:
```diff
@@ -0,0 +1,101 @@
+    var serviceDelegate: ServiceDelegate? { get set }
+
+    /// Is the service complete?
+    var isComplete: Bool { get }
```
Does "complete" mean properly configured, or?

darinkrauss replied:
Renamed to `hasValidConfiguration` to be more descriptive.

darinkrauss committed:
Refactor Analytics and RemoteData service protocols from review

- Refactor Analytics protocol to be more generic
- Remove unnecessary RemoteData protocol function
- Update ServiceDelegate protocol documentation
- Rename Service.isComplete to Service.hasValidConfiguration
- Rename Service.verify to Service.verifyConfiguration
- Move BatteryStatus, IOBStatus, and PumpStatus back to Rileylink
- Move LoopSettings back to Loop
- Move TempBasalRecommendation extension to LoopCore

darinkrauss committed:
Initial implementation of synchronizing remote data

darinkrauss commented:
Stale, closing in favor of new PR.
